### PR TITLE
fix/choose disagrements with file hierarchy

### LINKED ISSUE #1137: Where is "branch" set for a model component?
Hey all

There are FOCI-OpenIFS version 3.0 and 3.0.1. They are identical, except for one compile key in NEMO. Therefore, when I run `esm_master get-focioifs-3.0.1` ESM-Tools should pull NEMO from the same repo as for `3.0` but a different branch. 

In `nemo.yaml` we set
```python
choose_version: 
   ORCA05_LIM2_KCM_AOW_FS_OASISMCT4:
    requires:
    - nemobasemodel-3.6foci
    branch: esm-tools
    destination: nemo-${nemo.version}/CONFIG/${nemo.version}
    git-repository: https://git.geomar.de/foci/src/nemo_config/${nemo.version}.git
```

That's fine for FOCI-OpenIFS 3.0, but for 3.0.1 I want to pull branch "foci30". 
So in `focioifs.yaml` I set:
```python
nemo:
        choose_general.version:
                "3.0.1":
                        version: "ORCA05_LIM2_KCM_AOW_FS_OASISMCT4"
                        resolution: ORCA05
                        branch: foci30
                "3.0":
                        version: "ORCA05_LIM2_KCM_AOW_FS_OASISMCT4"
                        resolution: ORCA05
```

I would have expected that what is in `focioifs.yaml` overrides settings in `nemo.yaml` and thus switches from branch "esm_tools" to "foci30". 

But then this happens
```bash
(base) glogin7:/scratch/usr/shkjocke/model_test $ esm_master get-focioifs-3.0.1 -c 
    Executing commands in this order:
         mkdir -p focioifs-3.0.1
         cd focioifs-3.0.1
         git clone -b master https://git.geomar.de/foci/src/oasis3-mct5.git oasis
         git clone -b 2.5_r1910 https://git.geomar.de/foci/src/xios.git xios
         git clone -b foci22 https://gitlab.dkrz.de/ec-earth/oifs-43r3.git oifs-43r3-foci22
         chmod 700 -R oifs-43r3-foci22
         git clone -b esm-tools https://git.geomar.de/foci/src/nemo.git nemo-ORCA05_LIM2_KCM_AOW_FS_OASISMCT4
         git clone -b esm-tools https://git.geomar.de/foci/src/nemo_config/ORCA05_LIM2_KCM_AOW_FS_OASISMCT4.git nemo-ORCA05_LIM2_KCM_AOW_FS_OASISMCT4/CONFIG/ORCA05_LIM2_KCM_AOW_FS_OASISMCT4
         git clone -b foci211 https://gitlab.dkrz.de/ec-earth/runoff-mapper.git rnfmap-foci211
         sed -i '/COUPLENEMOECE = /s/.TRUE./.FALSE./g' oifs-43r3-foci22/src/ifs/module/yommcc.F90
         sed -i '/COUPLEFESOM2 = /s/.TRUE./.FALSE./g' oifs-43r3-foci22/src/ifs/module/yommcc.F90
         sed -i '/COUPLENEMOFOCI = /s/.FALSE./.TRUE./g' oifs-43r3-foci22/src/ifs/module/yommcc.F90
         cd ..
esm_master: check mode is activated. Not executing the actions above
```

As you can see, ESM-Tools tries to pull the `esm-tools` branch of NEMO (the one set in`nemo.yaml`) and not `foci30` (set in `focioifs.yaml`). 
I'm really puzzled by this. 
Every version of FOCI-OpenIFS uses the same repo for OpenIFS, but different branches. If I set in `focioifs.yaml` 
```python
oifs:
        choose_general.version:
                "3.0.1":
                        branch: test
```
it will try to pull the branch `test` for 3.0.1. 

Does anyone have any ideas why setting a branch in `focioifs.yaml` does not override what is in `nemo.yaml`? 

Best wishes
Joakim

### LINKED ISSUE #1221: Functionality to change component variables from setups in esm_master (part 2)
I found recently, that on release I can not overwrite a comp_command that is given in a choose block in the component, with another comp command that is given in a choose block in a setup. This seems related to: #32 

For FESOM we did a workaround by using the default command in the component, thereby allowing the setup comp command to overwrite.

For OpenIFS 48r1 I do not see how to make this work. 43r3 and 48r1 have too different comp commands in the component to use a single default.

### PR DESCRIPTION
**Pull request by mandresm**
Long standing issue, that might have been solve once for all, both for esm_runscripts and esm_master.

Blocks any `__setitem__` of values coming from lower hierarchy file category. For example, if ESM-Tools is trying to reassign a value coming from a component into a key that comes from a setup, it won't reassign it. This could be use to seamlessly `update` and change the `config` in the parser without having to care about the order, which will simplify the parser.

The problem: provenance takes ~1 second, so it is time intensive and we would become fully dependent on the provenance feature for the behavior of the parser... a bit dangerous.

**TODO**
- [x] Check differences in tests and see if they make sense
- [x] Clean up code
- [x] Run real tests in Levante
- [x] Check if  #1137 is solved

Closes #1221 (and a few more I guess...)
Closes  #1137

### CODE REVIEW

**Comment by mandresm** [comment]
This PR also solves problems reported by @PengyangSong and @ackerlar where the `echam.nproca: 24` and `echam.nprocb=24` variables, selected in `awiesm.yaml` where not overwritting the ones from the `echam.yaml` with values 16. You can see now that the tests are showing that this is changed and fixed: https://github.com/esm-tools/esm_tools/actions/runs/11165783421/job/31038282687?pr=1224#step:8:1293

**Comment by mandresm** [comment]
@pgierz, it is ready to review and merge. All tests run successfully, and differences are okay.

File: src/esm_master/compile_info.py [hunk 1]
```diff
@@ -336,7 +336,24 @@ def get_relevant_info(relevant_entries, raw_config, merge_into_this_config=None)
 
 class setup_and_model_infos:
     def __init__(self, vcs, general, parsed_args):
+        """
+        Initializes the setup and model information. It combines the components,
```
Comment (pgierz) [inline src/esm_master/compile_info.py:340]: Yay documentation

File: src/esm_parser/provenance.py [hunk 2]
```diff
@@ -511,7 +527,9 @@ def __setitem__(self, key, val):
         """
         Any time an item in a DictWithProvenance is set, extend the old provenance of
         the old value with the provenance of the new ``val`` and make that be the new
-        extended provenance history of the value.
+        extended provenance history of the value. Does not set the new valeu if the old
```
Comment (pgierz) [inline src/esm_parser/provenance.py]: Typo in the docs

File: src/esm_parser/provenance.py [hunk 3]
```diff
@@ -37,9 +37,24 @@
...
 
+CATEGORY_HIERARCHY = [
+    "defaults",  # Lowest in the hierarchy
+    "other_software",
+    "machines",
+    "components",
+    "setups",
+    "couplings",
+    "runscript",
+    "none",  # Highest in the hierarchy (no category means it is a change
```
Comment (pgierz) [inline src/esm_parser/provenance.py]: Here, and below, you can use None directly. I don't think you need a string. If you want a name, something like "set-by-function" (and maybe some logic to add the function name?)

**Review by pgierz — APPROVED** [review]
Looks good. I have one small comment suggestion for improvement, but it's not critical, this could go in as is. Nice work 🙂

File: src/esm_parser/provenance.py (hunk 2 above)
Comment (mandresm) [inline src/esm_parser/provenance.py]:
```suggestion
        extended provenance history of the value. Does not set the new value if the old
```

File: src/esm_parser/provenance.py (hunk 3 above)
Comment (mandresm) [reply src/esm_parser/provenance.py]: I have changed it with `backend`. The logic to add the function name already exist in the different methods of the `Provenance` class, but this does not belong to the `category` key of the provenance, but on the `modified_by` key.

**Comment by mandresm** [comment]
#approve-changes

**Comment by mandresm** [comment]
#bump

**Pull request merged**